node/src/reactor/main_reactor/control.rs

```diff
@@ -471,16 +484,30 @@ impl MainReactor {
     }
 
     pub(super) fn should_commit_upgrade(&self) -> bool {
-        let highest_switch_block_header = match &self.switch_block_header {
-            None => {
+        // header of latest complete block, and that block needs to be switch block
+        let highest_switch_block_header = match self.storage.read_highest_complete_block() {
+            Ok(Some(highest_complete_block)) => {
+                if highest_complete_block.header().is_switch_block() {
+                    highest_complete_block.take_header()
+                } else {
+                    return false;
+                }
+            }
+            Ok(None) => {
+                return false;
+            }
```

Comment:
```suggestion
            Ok(Some(highest_complete_block)) if highest_complete_block.header().is_switch_block( => {
                highest_complete_block.take_header()
            }
            Ok(Some(_)) | Ok(None) => {
                return false;
            }
```

Reply:
Fixed in https://github.com/casper-network/casper-node/pull/3971/commits/18bce1adea6345647e1568f2068b7ddc74c5ed1f